### ISSUE TITLE
Rejecting streamed data

### ISSUE THREAD

**Issue by rafaberaldo** [sveltejs/kit#9785]
### Describe the bug

When rejecting promises for streamed data, dev server usually crashes, and when it doesn't crash it doesn't show my custom error message.

### Reproduction

https://stackblitz.com/edit/sveltejs-kit-template-default-2dz1kb?file=src/routes/+page.server.js

### Logs

```Shell
[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Not Found".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```

### System Info

```Shell
System:
    OS: Linux 5.19 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (12) x64 AMD Ryzen 5 5600X 6-Core Processor
    Memory: 7.30 GB / 15.53 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
  Browsers:
    Firefox: 112.0.1
  npmPackages:
    @sveltejs/adapter-node: ^1.2.2 => 1.2.3 
    @sveltejs/kit: ^1.15.7 => 1.15.7 
    svelte: ^3.58.0 => 3.58.0 
    vite: ^4.3.1 => 4.3.1
```

### Severity

serious, but I can work around it

**Comment by gtm-nayan** [comment]
Cannot reproduce the crash with the provided repro on either stackblitz or locally, but can confirm the error reporting seems to be broken. It's logging undefined every time.

**Comment by rafaberaldo** [comment]
> Although I guess the expectation here is that you want "Custom error" to show up in the page as the error message, is that correct?
Yes

> For that you'd have to reject promises with the `error` helper exported from @sveltejs/kit otherwise it'll get treated as an unexpected error and only send `Internal Error` in the response.

OK I got it working, the syntax should be: ``reject(error(404, 'Custom error'));``

Thank you very much!

**Comment by ghostdevv** [comment]
Happens for me too

**Comment by adrifer** [comment]
Yes I have the same issue. It breaks my refresh auth logic, when I do a request inside a load function that returns a 401, I throw a redirect to /refresh/login. That works when I'm waiting inside the load or returning as a first level in the return object, but as soon as I move that request on a deeper level to stream the response, it crash the dev server

**Comment by gtm-nayan** [comment]
@adrifer Before you tread down that path, I should let you know that you can't throw redirects from a streamed promise. The server needs to know the status code before it starts sending the response and it's too late by the time it gets to your streamed promises.

As for the issue in general, Node.js requires that an error handler must be attached before the tick in which the promise is rejected. In SvelteKit, that's done by the serializer, but by the time we're done awaiting all [the load functions](https://github.com/sveltejs/kit/blob/73a91c644ed9f797b97f85d6f99b5b0cdc0f857f/packages/kit/src/runtime/server/page/index.js#L201-L202) it may have been too late to attach the handler.

I can't think of a clean way to solve this yet, I don't think it's actually possible with the current API, we'd have to walk the returned objects (arrays, maps, whatever included) as soon as we get it from the user but turns out even that's not enough.

Consider
```js
{ nested: { promise: Promise.reject() }, something: await setTimeout(1000) }
```
the rejected promise is created before the yield point of the timeout so the exception is triggered  and the process exits before the promise even reaches SvelteKit.

Perhaps there is some way to kludge this, ensuring we attach the error handlers before any other yield point is reached. The handlers would catch the error and put it back on the promise under some special key and check the presence of the key during serialization but I won't be too hopeful that it's viable. Perhaps we do need something like `defer()` after all.

**Comment by jetaggart** [comment]
FWIW, I am experiencing this and it hard crashes the prod serve as well. For example if I have a fetch failure it will hard crash and terminate the process:

For example if I have a fetch like this in a `+page.ts` **and** there's no server running on port 8080

```typescript
export function load(ctx) {
  return {
    streamed: {
      thing: ctx.fetch("http://localhost:8080")
    }
  };
}
```

The development server hard crashes:
```bash
> npm run build && npm run preview
```
<img width="667" alt="Screenshot 2023-05-12 at 5 25 33 PM" src="https://github.com/sveltejs/kit/assets/161737/27810863-cd48-4a7d-8b5f-544683f76b5c">

**Comment by jetaggart** [comment]
Here is a full repro: https://stackblitz.com/edit/sveltejs-kit-template-default-uq1e5c?file=src%2Froutes%2F%2Bpage.server.js

**Comment by MLNW** [comment]
This issue makes streamed data mostly unusable. Especially when dealing with fetches that may return a 404 or those that want to throw a redirect. It, for example, prevents us from using redirects to protect authentication sensitive routes.

And whats even worse is that the [documentation](https://kit.svelte.dev/docs/load#streaming-with-promises) of streamed data is suggesting that error handling isn't an issue. The example promise won't fail
```javascript
            three: new Promise((fulfil) => {
                setTimeout(() => {
                    fulfil(3)
                }, 1000);
            })
```
and the error handling
```js
    {#await data.streamed.three}
        Loading...
    {:then value}
        {value}
    {:catch error}
        {error.message}
    {/await}
```
suggests that rejected, streamed data does not crash the server.

Debugging this issue has cost us a lot of time that could have been prevent by mentioning this caveat in the documentation, or even better, implementing a fix. This should be a high priority issue, if you ask me.

**Comment by jetaggart** [comment]
Agreed with @MLNW. I ran into this issue specifically with trying to throw a redirect (which is the recommended approach) if the fetch failed on an unauthorized error. Right now we'd have to wrap this in something that is guaranteed not to fail and handle explicitly in our components.

**Comment by mpiorowski** [comment]
Also the same problem, the funny thing is that the 
```
{#await data.streamed.three}
        Loading...
    {:then value}
        {value}
    {:catch error}
        {error.message}
    {/await}
```
work when i navigate to this page from another route, BUT crash the server when i reload the page.

**Comment by knulpi** [comment]
> works when i navigate to this page from another route, BUT crash the server when i reload the page.

Could it be that you have a universal instead of a server-only load function set up? Because if it is universal, the client might handle it, and the error case here happens during SSR.

**Comment by knulpi** [comment]
The only solution for me right now to this, while still making use of the streaming feature is to catch the error in-place and handle the object in the page. This is not the same as classical error handling with load functions, but at least I can use `{#await}` blocks again to show loading state to the user.

```ts
return {
	promises: {
		sectionData: GetPageData(params.pageId).then(undefined, () => {return {error: "asdf"}})
	}
}
```

**Comment by eliasdefaria** [comment]
Still struggling with this as well. Does not seem like there's a viable solution at the moment other than to just handle the errors on the client. For redirects, I'm just check if I'm in the browser then using `goto`. Doesn't feel right :/

**Comment by munzx** [comment]
Had the same issue and since I could't find an answer I had to review the code. Turns out the error handle is actually a redirect object and the comment says:

"Create a `Redirect` object. If thrown during request handling, SvelteKit will return a redirect response.* Make sure you're not catching the thrown redirect, which would prevent SvelteKit from handling it."

While this has nothing to do with catching streaming error but I figured it could mean that with streaming we should as well let Sveltekit handles the error , and thus when catching an error all we need is to return the error text message and guess what , it worked. now am not sure how to handle the error status code but for now this works for me.

**Comment by munzx** [comment]
this is what I am using to resolve this issue 

`function streamedError(err: {status: number, body: {message: string}}) {
    return new Promise((fulfil, reject) => {
        setTimeout(() => {
            reject(error(err.status, err.body.message));
        }, 500);
    });
}`

and then I attach it to the streamed promises as follows:
`streamedFunc().catch(err=>streamedError(err))`

**Comment by jdgamble555** [comment]
Yup, this is definitely a problem.  If the `catch` on the frontend doesn't work, this is useless.  I don't think the docs need to get updated as much as this is a bug that needs fixed asap.  As said above, this makes streaming unusable.

J

**Comment by gtm-nayan** [comment]
We'll probably not be able to fix this without changing a few things in a major version because your promises can reject before they even reach SvelteKit.

In the meantime you can use workarounds https://jakearchibald.com/2023/unhandled-rejections/ to prevent crashing your server. Or set up a global promise rejection handler, which would be a good idea regardless of streamed promises.

For the ones that want to throw redirects from their streamed promises, that will not work, ever, because the browser has already received an `OK` response with the rest of the page by that point. The best you can do there is make another navigation on the client side like @eliasdefaria said.

**Comment by oscarhermoso** [comment]
Adding this snippet to the bottom of `svelte.config.js` will prevent crashes and help with debugging info during development

```
process.on('unhandledRejection', (reason, promise) => {
	console.error('Unhandled Rejection at: Promise', promise, 'reason:', reason);
});
```

**Comment by ddxv** [comment]
> We'll probably not be able to fix this without changing a few things in a major version because your promises can reject before they even reach SvelteKit.
> 
> In the meantime you can use workarounds https://jakearchibald.com/2023/unhandled-rejections/ to prevent crashing your server. Or set up a global promise rejection handler, which would be a good idea regardless of streamed promises.
> 
> For the ones that want to throw redirects from their streamed promises, that will not work, ever, because the browser has already received an `OK` response with the rest of the page by that point. The best you can do there is make another navigation on the client side like @eliasdefaria said.

Thank you for the workaround, but I am here from my first Svelte (and JS) project, and that seems a bit complex for me.

Is there another less complex way to get `{#await}` ... `{#catch}` working with an API that returns potential 404? I'm only trying to use Streaming Promises like in the docs for an example project. It seems to me the answer right now is that using Streaming Promises needs to wait for a fix?

**Comment by gtarsia** [comment]
@gtm-nayan is there a solution for this in svelte 5?
This comment is still true, https://github.com/sveltejs/kit/issues/9785#issuecomment-1552836871, the documentation is still misleading.

**Comment by gtarsia** [comment]
Here's my workaround: return data and error+code as normal data (like zod's `safeParse` return) in server code, then transform it back in your .svelte file.

```typescript
// $lib/stream.ts
export type SafeStream<T> = { data: T } | { code: number; message: string }
export type StreamPromise<T> = Promise<SafeStream<T>>

export async function stream<T>(fn: () => Promise<T>): StreamPromise<T> {
	const { promise, resolve } = Promise.withResolvers<SafeStream<T>>()
	try {
		resolve({ data: await fn() })
	} catch (err) {
		const message = (err as { message?: string })?.message ?? 'Internal Error'
		const code = (err as { code?: number })?.code ?? 500
		resolve({ code, message })
	}
	promise.catch(() => {}) // out of tick UnhandledPromiseRejection hack
	return promise
}

class StreamError extends Error {
	code: number
	constructor(code: number, message: string) {
		super(message)
		this.code = code
	}
}

export async function unstream<T>(promise: StreamPromise<T>): Promise<T> {
	const { promise: p, resolve, reject } = Promise.withResolvers<T>()
	const result = await promise
	if ('code' in result) {
		reject(new StreamError(result.code, result.message))
	} else {
		resolve(result.data)
	}
	return p
}

// +layout.server.ts or +page.server.ts
import { stream } from '$lib/stream'

export const load: LayoutServerLoad = ({ locals }) => {
	return {
		apps: stream(() => myData()),
	}
}

// +page.svelte
<script lang="ts">
  let { data } = $props()
  const apps = unstream(data.apps)
</script>
  {#await data}
    Loading...
  {:then data}
    Here's data
  {:catch error}
    <p>error {error.message}</p>
  {/await}

```

It is not the most elegant, but it:
- is type safe
- solves the next tick unhandled rejection problem
- lets you control the errors

Definitely a WIP workaround, I'll be updating this as I see fit:
- Svelte has a class `HttpError`, not sure if it's exported. If it is, it can replace my `SvelteError` class.
- You can call `goto` inside the `unstream` function in case you get a 401, but I think I could make it so it throws redirect. This would allow redirection to happen in case we're not streaming (but that would render the `render` fn name useless).

Let me know if someone has suggestions to improve this.

**Comment by yaman-cruise** [comment]
~~I had the same issue and it took me hours to figure out that I haven't used the fetch function from the loader ~~
properties.~~

Still the same error sorry

**Comment by yaman-cruise** [comment]
SvelteKit can handle errors with streaming. You need to ensure two things: 

1. Use the `fetch` provided by the event, not the native one.
2. Call the `fetch` inside the load method. Even if you pass the fetch to a custom fetch function, it will result in an error.

This approach works and triggers the `{:catch}` block in your template:

```js
export const load: LayoutServerLoad = async ({ fetch }) => {
	const session = await event.locals.auth();
	return {
		demo: fetch('/api/v1/users', {
				method: 'GET'
			})
			.then((res) => {
				if (!res.ok) {
					error(res.status);
				} else {
					return res.json();
				}
			})
	};
};
```

This doesn't work because fetch is not called inside the load function:
```js
const fetchUsers = async ({ fetch }: ServerLoadEvent) => {
	const res = await fetch('/api/supply/v1/account', {
		method: 'GET'
	});
	if (!res.ok) {
		error(res.status);
	} else {
		return (await res.json()) as Account;
	}
};
export const load: LayoutServerLoad = async ({ fetch }) => {
	const session = await event.locals.auth();
	return {
		demo: fetchUsers(fetch)
	};
};
```

All of this is mentioned in the documentation: https://svelte.dev/docs/kit/load#Streaming-with-promises. I believe we can close this ticket.

**Comment by W2Wizard** [comment]
Bump, I spent *way* too much time trying to find a solution to this issue.
Only thanks @gtarsia 's solution was I able to finally get it to work.

Please document this somewhere or link to this issue!

**Comment by teemingc** [comment]
This seems to have been fixed since 2.37.1 . I can't reproduce the crash beyond that version
